File: src/0112.path-sum.112/solution.go
package pathsum112

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	var dfs func(node *TreeNode, sum int) bool
	dfs = func(node *TreeNode, sum int) bool {
		sum += node.Val
		if node.Left == nil && node.Right == nil {
			return sum == targetSum
		}
		if node.Left != nil {
			left := dfs(node.Left, sum)
			if left {
				return true
			}
		}

		if node.Right != nil {
			right := dfs(node.Right, sum)
			if right {
				return true
			}
		}

		return false
	}

	return dfs(root, 0)

}

File: src/0112.path-sum.112/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        return dfs(root, targetSum);
    }

    private boolean dfs(TreeNode root, int rest) {
        rest -= root.val;
        if (root.left == null && root.right == null) {
            return rest == 0;
        }
        if (root.left != null && dfs(root.left, rest)) {
            return true;
        }
        return root.right != null && dfs(root.right, rest);
    }
}